src/fuelTanks/CCPACSVessel.cpp

```diff
@@ -0,0 +1,700 @@
+                         GetUID() + "\") must be larger than the cylinder radius (" + std::to_string(cylinderRadius) +
+                         ")!");
+    }
+    if (r <= 0. || r >= cylinderRadius) {
+        throw CTiglError("The knuckle radius (" + std::to_string(r) + ") of vessel \"" + GetName() + "\" (uID=\"" +
+                         GetUID() + "\") must be larger than 0 and smaller than the cylinder Radius (" +
+                         std::to_string(cylinderRadius) + ")!");
+    }
+
+    double c     = cylinderRadius - r; // Distance from the center to the center of the torus tube
+    double h     = R - sqrt((r + c - R) * (r - c - R)); // Height from the base of the dome to the top
+    double rt    = c * (1.0 + pow((R / r - 1), -1.0)); // Radius where the transition from sphere to torus occurs
+    double alpha = acos((rt - c) / r); // Angle under which the transition from sphere to torus occurs
+
+    gp_Dir dir(0.0, 1.0, 0.0);
+
+    gp_Pnt p1(R, 0.0, 0.0);
+    gp_Pnt p2(h, 0.0, cylinderRadius - r);
+
+    gp_Circ circ1(gp_Ax2(p1, dir), R);
+    gp_Circ circ2(gp_Ax2(p2, dir), r);
+
+    GC_MakeArcOfCircle arc1(circ1, -M_PI / 2, -alpha, Standard_True);
+    GC_MakeArcOfCircle arc2(circ2, -alpha, 0, Standard_True);
+
+    TopoDS_Edge domeEdge1 = BRepBuilderAPI_MakeEdge(arc1.Value()).Edge();
+    TopoDS_Edge domeEdge2 = BRepBuilderAPI_MakeEdge(arc2.Value()).Edge();
+
+    std::vector<TopoDS_Edge> edges = {domeEdge1, domeEdge2};
+
+    if (cylinderLength > 0.0) {
+        TopoDS_Vertex v1 = GetLastVertex(domeEdge2);
+        TopoDS_Vertex v2 = BRepBuilderAPI_MakeVertex(gp_Pnt(0.5 * cylinderLength + h, 0.0, cylinderRadius));
+
+        TopoDS_Edge cylinderEdge = BRepBuilderAPI_MakeEdge(v1, v2).Edge();
+        edges.push_back(cylinderEdge);
+    }
+
+    BuildVesselWire(edges, wire);
+}
+
+void CCPACSVessel::BuildVesselWireIsotensoid(BRepBuilderAPI_MakeWire& wire) const
+{
+    double cylinderRadius = m_cylinderRadius_choice2.get();
+    double cylinderLength = m_cylinderLength_choice2.get();
+
+    double polarOpeningRadius = _isotensoidPtr->GetPolarOpeningRadius();
+    if (polarOpeningRadius <= 0 || polarOpeningRadius >= cylinderRadius) {
+        throw CTiglError("The polar opening radius (" + std::to_string(polarOpeningRadius) + ") of vessel \"" +
+                         GetName() + "\" (uID=\"" + GetUID() +
+                         "\") must be larger than 0 and smaller than the cylinder radius (" +
+                         std::to_string(cylinderRadius) + ")!");
+    }
+
+    std::vector<double> x, r;
+    IsotensoidContour(cylinderRadius, polarOpeningRadius, 50, x, r);
+    double h = x.back() - x.front();
+
+    TColgp_Array1OfPnt array(1, x.size());
+    for (size_t i = 0; i < x.size(); ++i) {
+        array.SetValue(i + 1, gp_Pnt(-x[i], 0.0, r[i]));
+    }
+    Handle(Geom_BSplineCurve) bspline = GeomAPI_PointsToBSpline(array).Curve();
+    TopoDS_Edge domeEdge              = BRepBuilderAPI_MakeEdge(bspline);
```

Comment:
The B-Spline approximation could move to the IsotensoidContour struct as well (see above)

Reply:
Right, moved it into the struct (c427b08e).